### ISSUE TITLE
run full_sort_topk, sometimes ok and sometimes throw such follow exceptions

### ISSUE THREAD

**Issue by dxjjhm** [RUCAIBox/RecBole#848]
topk_score, topk_iid_list = full_sort_topk(np.asarray(uid_series), model, test_data,  topk)  

Traceback (most recent call last):
  File "quick_load_and_predict.py", line 195, in load_model_predict_1by1
    topk_score, topk_iid_list = full_sort_topk(np.asarray(uid_series), model, test_data, args.topk)
  File "/data/home/anaconda3/envs/recbole0.2.1/lib/python3.7/site-packages/recbole-0.2.1-py3.7.egg/recbole/utils/case_study.py", line 87, in full_sort_topk
    scores = full_sort_scores(uid_series, model, test_data)
  File "/data/home/anaconda3/envs/recbole0.2.1/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 26, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/anaconda3/envs/recbole0.2.1/lib/python3.7/site-packages/recbole-0.2.1-py3.7.egg/recbole/utils/case_study.py", line 45, in full_sort_scores
    history_row = torch.cat([torch.full_like(hist_iid, i) for i, hist_iid in enumerate(history_item)])
RuntimeError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors), but no fallback function is registered for schema aten::_cat.  This usually means that this function requires a non-empty list of Tensors.  Available functions are [CPU, CUDA, QuantizedCPU, BackendSelect, Named, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, Autocast, Batched, VmapMode].

CPU: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/build/aten/src/ATen/CPUType.cpp:2127 [kernel]
CUDA: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/build/aten/src/ATen/CUDAType.cpp:2983 [kernel]
QuantizedCPU: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/build/aten/src/ATen/QuantizedCPUType.cpp:297 [kernel]
BackendSelect: fallthrough registered at /opt/conda/conda-bld/pytorch_1607370141920/work/aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
Named: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/aten/src/ATen/core/NamedRegistrations.cpp:7 [backend fallback]
AutogradOther: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradCPU: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradCUDA: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradXLA: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradPrivateUse1: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradPrivateUse2: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
AutogradPrivateUse3: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/VariableType_2.cpp:8078 [autograd kernel]
Tracer: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/torch/csrc/autograd/generated/TraceType_2.cpp:9654 [kernel]
Autocast: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/aten/src/ATen/autocast_mode.cpp:258 [kernel]
Batched: registered at /opt/conda/conda-bld/pytorch_1607370141920/work/aten/src/ATen/BatchingRegistrations.cpp:511 [backend fallback]
VmapMode: fallthrough registered at /opt/conda/conda-bld/pytorch_1607370141920/work/aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]

**Comment by 2017pxy** [comment]
@dxjjhm Hi, from the error message I guess this bug is due to the lack of history interaction, which means some users in `uid_series` don't have interaction records in training set. My suggestion is that you can check if your `uid_series` is correct. 

By the way, `0` is the padding uid in our framework (if your dataset is not loaded via `benchmark_filename`), so your `uid_series` should not include `0`.

**Comment by dxjjhm** [comment]
In my dataset, uid and iid  min value is 1, but after training then predict, always some user throw exception as above, that means those user have interaction but no recommendation.

train, validation, test data split ratio as 8:1:1, does you mean those error predict users in the validation dataset or test dataset ?

**Comment by 2017pxy** [comment]
@dxjjhm Hi, since the information you given us is limit and vague (without any message about data and your `uid_series`), I am not entirely sure what the problem is. 

As I have said before, according to the error message,  I think this problem is due to the lack of history interaction. Since you said this problem only arises for some certain users, why not make some simple tests to check out this problem by yourself? I believe you can figure out the problem by comparing the difference between the error users and the normal users.

About your question:
>  train, validation, test data split ratio as 8:1:1, does you mean those error predict users in the validation dataset or test dataset ?

In fact, if the `group_by_user` in your config is `True`, the data will be splitted by user and each user will generally occur in all three sets (train, valid and test sets). 

Let me take an example: 
If the split ratio is [0.8, 0.1, 0.1], and the original data is like this (NOTE: it is not the atomic file, just an example):
```
UserID	ItemID_list
1	[1,2,3,4,5,6,7,8,9,10]
2	[1,2,3,4,5,6,7,8,9,10]
3	[1,2,3,4,5,6,7,8,9,10]
4	[1,2,3,4,5,6,7,8,9,10]
5	[1,2,3,4,5,6,7,8,9,10]
```   
In RecBole, data will be splitted as:
```
Train:
1	[1,2,3,4,5,6,7,8]
2	[1,2,3,4,5,6,7,8]
3	[1,2,3,4,5,6,7,8]
4	[1,2,3,4,5,6,7,8]
5	[1,2,3,4,5,6,7,8]
Valid:
1	[9]
2	[9]
3	[9]
4	[9]
5	[9]
Test:
1	[10]
2	[10]
3	[10]
4	[10]
5	[10]
```
And I hope you can understand the data splitting in RecBole according to my explanation. Thanks.

**Comment by dxjjhm** [comment]
@2017pxy  I am really grateful for your answer. I will check those expection users data.
And my code paste as follow
`
def load_model_predict_1by1(args):
    """
    # 加载模型并预测推荐结果
    """
    model_name = args.model
    dataset_name = args.dataset
    seq = args.seq
    prop_path = args.prop_path
    checkpoint_dir = "{}/general/saved/{}/{}/{}".format(prop_path, dataset_name, model_name, seq)
    params = {
        "gpu_id": args.gpu_id,
        "use_gpu": args.use_gpu,
        'checkpoint_dir': checkpoint_dir,
        "data_path": args.data_path
    }

    # 需要设置负采样为0的算法，即不需要进行负采样
    register_table = {
        "MultiDAE": 1,
        "MultiVAE": 1,
        'MacridVAE': 1,
        'CDAE': 1,
        'ENMF': 1,
        'RaCT': 1,
        'RecVAE': 1
    }

    if model_name in register_table:
        params.update({"training_neg_sample_num": 0})

    print(params)
    yaml_path = "{}/properties/{}.{}.{}.yaml".format(prop_path, model_name, dataset_name, seq)
    config = Config(model=model_name, dataset=dataset_name, config_file_list=[yaml_path], config_dict=params)
    init_seed(config['seed'], config['reproducibility'])
    print(config)

    dataset_path = os.path.join(config['checkpoint_dir'], '{}-dataset.pth'.format(dataset_name))
    print(dataset_path)
    with open(dataset_path, 'rb') as f:  # You can use your filtered data path here.
        dataset = pickle.load(f)

    dataloader_path = os.path.join(config['checkpoint_dir'], '{}-for-{}-dataloader.pth'.format(dataset_name, model_name))
    print(dataloader_path)
    train_data, valid_data, test_data = load_split_dataloaders(dataloader_path)
    print(train_data.dataset)
    print(valid_data.dataset)
    print(test_data.dataset)

    # test_data.user_df.interaction[dataset.uid_field].to(config['device'])
    # model = get_model(config['model'])(config, train_data).to(config['device'])
    model = get_model(config['model'])(config, train_data)

    model_path = os.path.join(config['checkpoint_dir'], args.model_saved_name)  # 训练完成后保存的模型路径 xxx/xxx/xxx.pth
    print(model_path)
    checkpoint = torch.load(model_path)  # Here you can replace it by your model path.
    model.load_state_dict(checkpoint['state_dict'])
    print(model)

    # 按批次循环预计所有用户的推荐结果
    user_num = dataset.user_num  # 获取全部的用户个数
    uid_tmp = 0

    f_res = pd.DataFrame(columns=[dataset.uid_field, dataset.iid_field, "score"])
    while uid_tmp <= user_num:
        try:
            uid_series = dataset.token2id(dataset.uid_field, [str(uid_tmp)])  # 原始数据集中的用户id，变换为训练内部使用的索引id
            topk_score, topk_iid_list = full_sort_topk(np.asarray(uid_series), model, test_data, args.topk)
            # print(topk_score.shape)
            [l, w] = topk_score.shape
            flat_uid_series = np.repeat(uid_series, w, axis=0)  # 这里没有使用topk 而是 w, 即要求和输出的结果一致,
            uid_df = DataFrame(flat_uid_series, columns=[dataset.uid_field], dtype=int)

            # 返回的topk的物品id是训练内部使用的索引id, 转化为原始数据的物品id
            external_item_list = dataset.id2token(dataset.iid_field, topk_iid_list)
            flat_items = external_item_list.reshape(1, l * w)
            rec_item_df = DataFrame(flat_items.transpose(), columns=[dataset.iid_field], dtype=int)
            flat_score = topk_score.numpy().reshape(1, l * w)
            rec_score_df = DataFrame(flat_score.transpose(), columns=['score'])
            cat = pd.concat([uid_df, rec_item_df, rec_score_df], axis=1)
            f_res = pd.concat([f_res, cat], axis=0)
        except ValueError as v:
            print("uid = {} not exist in dataset".format(uid_tmp))
            # print(str(v))
            # traceback.print_exc()
        except RuntimeError as e:
            print("uid = {} run time error".format(uid_tmp))
            # print("predict run time error")
            # print(str(e))
            # traceback.print_exc()
        finally:
            uid_tmp += 1

    cur_time = time.strftime('%Y%m%d%H%M%S', time.localtime(time.time()))
    save_file = os.path.join(checkpoint_dir, f"{model_name}_predict_{cur_time}.txt")
    f_res.to_csv(save_file, index=False, header=True)
`

**Comment by dxjjhm** [comment]
uid = 135882 not exist in dataset
uid = 135883 run time error, has 5 num interactions
uid = 135885 run time error, has 5 num interactions
uid = 135886 run time error, has 6 num interactions
uid = 135887 run time error, has 5 num interactions
uid = 135888 run time error, has 4 num interactions
uid = 135889 run time error, has 4 num interactions
uid = 135890 run time error, has 6 num interactions


By the way, when the user has  less then 10 interactions,  and split ratio as [ 0.8, 01, 01]  it will run fails.
@2017pxy, does this conclusion is right?

**Comment by 2017pxy** [comment]
@dxjjhm  Hi,  I think you figure it out! 

The ratio splitting for users who have a small number of  interactions may probably casue some strange problems, like some users may not have testing set.  In fact, for this problem, we have already updated a fix in #787 by adding some speacial checking to avoid some unexpected situations, although the solution may not prefect. 

Since the fix is updated recently, it is only merged into our master branch but not in our v0.2.1 Pypi package. So I advise you to check the version of your code, pull the lastest code from our master branch and try again to see if the problem is fixed. If not, I am looking forward to your feedback.

Thanks!
